mindsdb/interfaces/model/model_controller.py

```diff
@@ -1,366 +1,383 @@
-        return predictor
-    
-    def learn_for_update(self, name, from_data, to_predict, datasource_id, kwargs={}, company_id=None):
-        kwargs['join_learn_process'] = True
-        return self.learn(name, from_data, to_predict, datasource_id, kwargs, company_id, False)
-
-    def learn(self, name, from_data, to_predict, datasource_id, kwargs={}, company_id=None, save=True):
-        from mindsdb.interfaces.model.learn_process import LearnProcess, run_learn
-
+    def learn(self, name: str, from_data: dict, to_predict: str, datasource_id: int, kwargs: dict, company_id: int) -> None:
         create_process_mark('learn')
-        original_name = name
-        name = f'{company_id}@@@@@{name}'
+
+        problem_definition = {'target': to_predict if isinstance(to_predict, str) else to_predict[0]}
         join_learn_process = kwargs.get('join_learn_process', False)
-        
-        if save:
-            self._setup_for_creation(name, original_name, company_id=company_id)
-
-        if self.ray_based:
-            run_learn(
-                name=name,
-                db_name=original_name,
-                from_data=from_data,
-                to_predict=to_predict,
-                kwargs=kwargs,
-                datasource_id=datasource_id,
-                company_id=company_id,
-                save=save
-            )
+        if 'join_learn_process' in kwargs:
+            del kwargs['join_learn_process']
 
-        else:
-            p = LearnProcess(name, original_name, from_data, to_predict, kwargs, datasource_id, company_id, save)
-            p.start()
-            if join_learn_process is True:
-                p.join()
-                if p.exitcode != 0:
-                    delete_process_mark('learn')
-                    raise Exception('Learning process failed !')
+        # Adapt kwargs to problem definition
+        if 'timeseries_settings' in kwargs:
+            problem_definition['timeseries_settings'] = kwargs['timeseries_settings']
 
-        delete_process_mark('learn')
-        return 0
+        if 'stop_training_in_x_seconds' in kwargs:
+            problem_definition['stop_after'] = kwargs['stop_training_in_x_seconds']
 
-    def predict(self, name, pred_format, when_data=None, kwargs={}, company_id=None):
-        from mindsdb_datasources import (FileDS, ClickhouseDS, MariaDS,
-                                         MySqlDS, PostgresDS, MSSQLDS, MongoDS,
-                                         SnowflakeDS, AthenaDS)
-        import mindsdb_native
-        from mindsdb.interfaces.storage.db import session, Predictor
+        self.generate_predictor(name, from_data, datasource_id, problem_definition, company_id)
 
+        # TODO: Should we support kwargs['join_learn_process'](?)
+        self.fit_predictor(name, from_data, join_learn_process, company_id)
+
+    def predict(self, name: str, when_data: dict, pred_format: str, company_id: int):
```

Comment:
```when_data: Union[dict, list]``` ?

Reply:
Should be `Union[dict, list, pd.DataFrame]` even, well spotted